### ISSUE TITLE
Monix Task.gather() blocks Http4s client pool from reclaiming failed tasks

### ISSUE THREAD

**Issue by ngbinh** [http4s/http4s#2251]
Using Task.sequence() works as expected.

The repro can be found here https://github.com/ngbinh/http4s-monix

snip of tests

the setup:

```scala
def test(traverse: Seq[Task[String]] => Task[Seq[String]], timeout: FiniteDuration): Future[Assertion] = {
    val httpScheduler: Scheduler = Scheduler.fixedPool(
      name = "http4s-monix-scheduler",
      poolSize = 4,
      executionModel = ExecutionModel.AlwaysAsyncExecution
    )

    val task = BlazeClientBuilder[Task](httpScheduler)
      .withMaxTotalConnections(maxTotalConnections = 4)
      .withRequestTimeout(FiniteDuration(20, TimeUnit.SECONDS))
      .resource
      .use { client =>
        val list: Seq[Task[String]] = List.range(1, 30).map { i =>
          val uri = Uri(
            scheme = Some(Scheme.https),
            authority = Some(Authority(host = RegName("httpbin.org"))),
            path = s"/status/${500 + i}"
          )
          val request = Request[Task](
            method = Method.GET,
            uri = uri
          )
          client.expect[String](request).onErrorHandleWith { error =>
            log.error(error)(s"request to $uri")
            Task.raiseError(error)
          }
        }

        for {
          _ <- traverse(list).onErrorFallbackTo(Task(Seq.empty[String]))
          _ <- traverse(list).onErrorFallbackTo(Task(Seq.empty[String]))
        } yield ()
      }

    task.timeout(timeout).attempt.runToFuture.map(_ should be ('right))
  }
```

the tests:
```scala
  "Monix" - {
    "should works with sequence()" in {
      test(Task.sequence, FiniteDuration(10, TimeUnit.SECONDS))
    }

    "but fails with gather()" in {
      test(Task.gather, FiniteDuration(10, TimeUnit.SECONDS))
     // this test is timed out 
    }
  }
```

**Comment by ngbinh** [comment]
Thanks for the quick reply.
```scala
[info] Http4sMonixSpec:
[info] Monix with AsyncHttpClient
[info] - should works with sequence()
[info] - still works with gather()
[info] - still works with gatherUnordered()
[info] Monix with Blaze
[info] - should works with sequence()
[info] - should works with gatherUnordered() *** FAILED ***
[info]   Left(java.util.concurrent.TimeoutException: Task timed-out after 10 seconds of inactivity) was not right (Http4sMonixSpec.scala:109)
[info] - but fails with gather() *** FAILED ***
[info]   Left(java.util.concurrent.TimeoutException: Task timed-out after 10 seconds of inactivity) was not right (Http4sMonixSpec.scala:109)
```

So it works well with AsyncHttpClient on all three cases. But failed on Blaze with `Task.gatherUnordered` and `Task.gather`. 

Will try on `parMapN` and `IO`

**Comment by rossabaker** [comment]
The connection pool is fully allocated, and no progress gets made beyond the first four requests.  The test passes when the status starts at 200, making the task succeed.  The pool manager needs to recycle the connection on error, and is not.

Going to see if I can write a failing unit test against IO, so we know it's not Monix specific (I am fairly certain it is not) and fix it.

**Comment by rossabaker** [comment]
It's not Monix.  I have a failing unit test.

**Comment by ngbinh** [comment]
looks like when there is an error, the connection isn't released at all. I pushed the tests to show the failure of even `Task.sequence` in https://github.com/ngbinh/http4s-monix

**Comment by ngbinh** [comment]
https://github.com/ngbinh/http4s-monix/tree/ver-0.18.21 for `0.18.21` and Blaze client works fine

**Comment by rossabaker** [comment]
I'm not sure I'm seeing the same thing as you on [this commit in http4s-monix](https://github.com/ngbinh/http4s-monix/commit/3e4dc362bbe87491a11e702a6142a893556f3862).  I don't see a failure on `sequence`, and async-http-client is also failing.  (Stack traces trimmed):

```
[info] Http4sMonixSpec:
[info] Monix with AsyncHttpClient
[info] - should works with sequence()
[info] - still works with gather() *** FAILED ***
[info]   Left(org.asynchttpclient.exception.TooManyConnectionsException: Too many connections: 4) was not right (Http4sMonixSpec.scala:99)
[info] - still works with gatherUnordered() *** FAILED ***
[info]   Left(org.asynchttpclient.exception.TooManyConnectionsException: Too many connections: 4) was not right (Http4sMonixSpec.scala:107)
[info] Monix with Blaze[Task]
[info] - should works with sequence()
[info] - should works with gather() *** FAILED ***
[info]   java.util.concurrent.TimeoutException: Task timed-out after 10 seconds of inactivity
[info] - but fails with gatherUnordered() *** FAILED ***
[info]   java.util.concurrent.TimeoutException: Task timed-out after 10 seconds of inactivity
```

**Comment by ngbinh** [comment]
@rossabaker yes. `AsyncHttpClient` also fails, but for a different reason. 

I saw `sequence()` sometimes fail, not always though. I guess it could just be the 10 seconds timeout is too short.

**Comment by ngbinh** [comment]
In essence, I think the connections were not released right away after failures, they can only be reclaimed after the explicitly set timeout.

**Comment by rossabaker** [comment]
Fixed by #2258.
